crates/wit-component/src/linking.rs

```diff
@@ -0,0 +1,1383 @@
+        if let Some((key, value)) = exporters.get_key_value(&export.key) {
+            match value.as_slice() {
+                [] => unreachable!(),
+                [exporter] => {
+                    resolved.insert(*key, *exporter);
+                }
+                _ => {
+                    duplicates.push((metadata.name, *key, value.as_slice()));
+                }
+            }
+        } else {
+            missing.push((metadata.name, export));
+        }
+    };
+
+    for metadata in metadata {
+        for (name, (ty, flags)) in &metadata.env_imports {
+            triage(
+                metadata,
+                Export {
+                    key: ExportKey {
+                        name,
+                        ty: Type::Function(ty.clone()),
+                    },
+                    flags: *flags,
+                },
+            );
+        }
+
+        for name in &metadata.memory_address_imports {
+            triage(
+                metadata,
+                Export {
+                    key: ExportKey {
+                        name,
+                        ty: Type::Global(GlobalType {
+                            ty: ValueType::I32,
+                            mutable: false,
+                        }),
+                    },
+                    flags: 0,
+                },
+            );
+        }
+    }
+
+    for metadata in metadata {
+        for name in &metadata.table_address_imports {
+            if let Some((key, value)) = function_exporters.get(name) {
+                match value.as_slice() {
+                    [] => unreachable!(),
+                    [exporter] => {
+                        resolved.insert(key, *exporter);
+                    }
+                    _ => {
+                        duplicates.push((metadata.name, *key, value.as_slice()));
+                    }
+                }
+            } else {
+                missing.push((
+                    metadata.name,
+                    Export {
+                        key: ExportKey {
+                            name,
+                            ty: Type::Function(FunctionType {
+                                parameters: Vec::new(),
+                                results: Vec::new(),
+                            }),
+                        },
+                        flags: 0,
+                    },
+                ));
+            }
+        }
+    }
+
+    (resolved, missing, duplicates)
+}
+
+/// Recursively add a library (represented by its offset) and its dependency to the specified set, maintaining
+/// topological order (modulo cycles).
+fn topo_add<'a>(
+    sorted: &mut IndexSet<usize>,
+    dependencies: &HashMap<usize, HashSet<usize>>,
+    element: usize,
+) {
+    let empty = &HashSet::new();
+    let deps = dependencies.get(&element).unwrap_or(empty);
+
+    // First, add any dependencies which do not depend on `element`
+    for &dep in deps {
```

Comment:
One thing to be careful of here is that `deps` is a `HashSet` which has non-deterministic iteration order between runs. This might be a good place to use `IndexSet` to ensure the output of the `link` command is deterministic.

Reply:
I've changed this to an `IndexSet`.